core/base/timer.cpp

```diff
@@ -0,0 +1,137 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/timer.hpp>
+
+
+#include <chrono>
+#include <memory>
+
+
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/executor.hpp>
+
+
+#include "cuda/base/device.hpp"
+#include "dpcpp/base/device.hpp"
+#include "hip/base/device.hpp"
+
+
+namespace gko {
+
+
+time_point::time_point() : type_{type::cpu}, data_{} {}
+
+
+time_point::~time_point()
+{
+    switch (type_) {
+    case type::cuda:
+        kernels::cuda::destroy_event(data_.cuda_event);
+        break;
+    case type::hip:
+        kernels::hip::destroy_event(data_.hip_event);
+        break;
+    case type::dpcpp:
+        kernels::dpcpp::destroy_event(data_.dpcpp_event);
+        break;
+    case type::cpu:
+    default:
+        break;
+    }
+}
+
+
+time_point::time_point(time_point&& other)
+    : type_{std::exchange(other.type_, type::cpu)}, data_{other.data_}
+{}
+
+
+time_point& time_point::operator=(time_point&& other)
+{
+    // make sure we release the data via other
+    std::swap(type_, other.type_);
+    std::swap(data_, other.data_);
+    return *this;
```

Comment:
```suggestion
  type_ = std::exchange(other.type_, type::cpu);
  data_ = std::exchange(other.data_, {nullptr});
```
Otherwise, the moved-from state might not be uniquely defined.

Reply:
this would leave the resources previously stored in `*this` dangling